File: Sources/Mobi2EpubTransferCore/TextEpubPostProcessor.swift
import Foundation

public struct TextEpubPostProcessResult: Equatable, Sendable {
    public let repairedIdentifierCount: Int
    public let removedBrokenResourceCount: Int
    public let reorderedNavigationCount: Int

    public var reportText: String {
        """
        Text EPUB post-processing
        -------------------------
        EPUB 3 semantic markup preserved: yes
        Invalid identifiers repaired: \(repairedIdentifierCount)
        Broken style/resource references removed: \(removedBrokenResourceCount)
        Navigation playOrder values repaired: \(reorderedNavigationCount)
        """
    }
}

public enum TextEpubPostProcessError: Error, Equatable {
    case unzipFailed(String)
    case zipFailed(String)
}

/// Repairs structural defects commonly carried from KF8/AZW3 sources without
/// rewriting book text or removing semantic elements such as `<ruby>`.
public struct TextEpubPostProcessor: Sendable {
    private let runner: any ProcessRunning

    public init(runner: any ProcessRunning = ProcessRunner()) {
        self.runner = runner
    }

    public func process(epubURL: URL) async throws -> TextEpubPostProcessResult {
        let fileManager = FileManager.default
        let workspace = fileManager.temporaryDirectory
            .appendingPathComponent("MobiVerseTextEpubRepair", isDirectory: true)
            .appendingPathComponent(UUID().uuidString, isDirectory: true)
        let extracted = workspace.appendingPathComponent("book", isDirectory: true)
        try fileManager.createDirectory(at: extracted, withIntermediateDirectories: true)
        defer { try? fileManager.removeItem(at: workspace) }

        let unzip = try await runner.run(
            executableURL: URL(fileURLWithPath: "/usr/bin/unzip"),
            arguments: ["-qq", epubURL.path, "-d", extracted.path]
        )
        guard unzip.exitCode == 0 else { throw TextEpubPostProcessError.unzipFailed(unzip.output) }

        let contentFiles = regularFiles(in: extracted)
        let repairedIdentifiers = try repairInvalidIdentifiers(in: contentFiles)
        let removedResources = try removeBrokenResourceReferences(in: contentFiles, root: extracted)
        let reorderedNavigation = try repairNavigationOrder(in: contentFiles)

        guard repairedIdentifiers + removedResources + reorderedNavigation > 0 else {
            return TextEpubPostProcessResult(
                repairedIdentifierCount: 0,
                removedBrokenResourceCount: 0,
                reorderedNavigationCount: 0
            )
        }

        let rebuiltURL = epubURL.deletingLastPathComponent()
            .appendingPathComponent(".mobiverse-text-repair-\(UUID().uuidString).epub")
        defer { try? fileManager.removeItem(at: rebuiltURL) }

        let mimetypeURL = extracted.appendingPathComponent("mimetype")
        if fileManager.fileExists(atPath: mimetypeURL.path) {
            let mimetypeZip = try await runner.run(
                executableURL: URL(fileURLWithPath: "/usr/bin/zip"),
                arguments: ["-X0q", rebuiltURL.path, "mimetype"],
                currentDirectoryURL: extracted
            )
            guard mimetypeZip.exitCode == 0 else { throw TextEpubPostProcessError.zipFailed(mimetypeZip.output) }
        }

        let rootEntries = try fileManager.contentsOfDirectory(atPath: extracted.path)
            .filter { $0 != "mimetype" }
            .sorted()
        if !rootEntries.isEmpty {
            let contentZip = try await runner.run(
                executableURL: URL(fileURLWithPath: "/usr/bin/zip"),
                arguments: ["-Xr9qD", rebuiltURL.path] + rootEntries,
                currentDirectoryURL: extracted
            )
            guard contentZip.exitCode == 0 else { throw TextEpubPostProcessError.zipFailed(contentZip.output) }
        }

        _ = try fileManager.replaceItemAt(epubURL, withItemAt: rebuiltURL)
        return TextEpubPostProcessResult(
            repairedIdentifierCount: repairedIdentifiers,
            removedBrokenResourceCount: removedResources,
            reorderedNavigationCount: reorderedNavigation
        )
    }

    private func regularFiles(in root: URL) -> [URL] {
        guard let enumerator = FileManager.default.enumerator(
            at: root,
            includingPropertiesForKeys: [.isRegularFileKey],
            options: [.skipsHiddenFiles]
        ) else { return [] }
        return enumerator.compactMap { item in
            guard let url = item as? URL,
                  (try? url.resourceValues(forKeys: [.isRegularFileKey]).isRegularFile) == true else { return nil }
            return url
        }
    }

    private func repairInvalidIdentifiers(in files: [URL]) throws -> Int {
        let extensions = Set(["xhtml", "html", "htm", "xml", "opf", "ncx"])
        let idRegex = try NSRegularExpression(pattern: #"\bid\s*=\s*([\"'])([^\"']+)\1"#, options: [.caseInsensitive])
        var mapping: [String: String] = [:]
        var used = Set<String>()

        for file in files where extensions.contains(file.pathExtension.lowercased()) {
            let text = try String(contentsOf: file, encoding: .utf8)
            let nsText = text as NSString
            for match in idRegex.matches(in: text, range: NSRange(location: 0, length: nsText.length)) {
                let identifier = nsText.substring(with: match.range(at: 2))
                used.insert(identifier)
                guard !isValidXMLIdentifier(identifier), mapping[identifier] == nil else { continue }
                var replacement = sanitizedIdentifier(identifier)
                var suffix = 2
                while used.contains(replacement) {
                    replacement = "\(sanitizedIdentifier(identifier))-\(suffix)"
                    suffix += 1
                }
                mapping[identifier] = replacement
                used.insert(replacement)
            }
        }

        guard !mapping.isEmpty else { return 0 }
        for file in files where extensions.contains(file.pathExtension.lowercased()) {
            var text = try String(contentsOf: file, encoding: .utf8)
            let original = text
            for (invalid, valid) in mapping {
                let escaped = NSRegularExpression.escapedPattern(for: invalid)
                text = text.replacingOccurrences(
                    of: #"(\bid\s*=\s*[\"'])"# + escaped + #"([\"'])"#,
                    with: "$1\(valid)$2",
                    options: [.regularExpression, .caseInsensitive]
                )
                text = text.replacingOccurrences(of: "#\(invalid)", with: "#\(valid)")
            }
            if text != original { try text.write(to: file, atomically: true, encoding: .utf8) }
        }
        return mapping.count
    }

    private func removeBrokenResourceReferences(in files: [URL], root: URL) throws -> Int {
        var removedCount = 0
        for file in files where file.pathExtension.lowercased() == "css" {
            var css = try String(contentsOf: file, encoding: .utf8)
            let original = css
            let fontFaceRegex = try NSRegularExpression(
                pattern: #"@font-face\s*\{.*?\}"#,
                options: [.caseInsensitive, .dotMatchesLineSeparators]
            )
            let nsCSS = css as NSString
            let blocks = fontFaceRegex.matches(in: css, range: NSRange(location: 0, length: nsCSS.length))
                .map { nsCSS.substring(with: $0.range) }
            for block in blocks where containsMissingURL(in: block, relativeTo: file, root: root) {
                css = css.replacingOccurrences(of: block, with: "")
                removedCount += 1
            }

            let missingURLs = missingURLTokens(in: css, relativeTo: file, root: root)
            for token in missingURLs {
                let escaped = NSRegularExpression.escapedPattern(for: token)
                let declaration = #"(?im)^\s*[-a-z0-9_]+\s*:\s*[^;{}]*"# + escaped + #"[^;{}]*;\s*"#
                let before = css
                css = css.replacingOccurrences(of: declaration, with: "", options: .regularExpression)
                if css != before { removedCount += 1 }
            }
            if css != original { try css.write(to: file, atomically: true, encoding: .utf8) }
        }

        let markupExtensions = Set(["xhtml", "html", "htm"])
        for file in files where markupExtensions.contains(file.pathExtension.lowercased()) {
            var markup = try String(contentsOf: file, encoding: .utf8)
            let original = markup
            let linkRegex = try NSRegularExpression(pattern: #"<link\b[^>]*>"#, options: [.caseInsensitive])
            let nsMarkup = markup as NSString
            let links = linkRegex.matches(in: markup, range: NSRange(location: 0, length: nsMarkup.length))
                .map { nsMarkup.substring(with: $0.range) }
            for link in links where link.range(of: "stylesheet", options: .caseInsensitive) != nil {
                guard let href = attribute(named: "href", in: link),
                      isMissingLocalResource(href, relativeTo: file, root: root) else { continue }
                markup = markup.replacingOccurrences(of: link, with: "")
                removedCount += 1
            }
            if markup != original { try markup.write(to: file, atomically: true, encoding: .utf8) }
        }
        return removedCount
    }

    private func repairNavigationOrder(in files: [URL]) throws -> Int {
        let playOrderRegex = try NSRegularExpression(pattern: #"playOrder\s*=\s*([\"'])\d+\1"#, options: [.caseInsensitive])
        var repaired = 0
        for file in files where file.pathExtension.lowercased() == "ncx" {
            var text = try String(contentsOf: file, encoding: .utf8)
            let nsText = text as NSString
            let matches = playOrderRegex.matches(in: text, range: NSRange(location: 0, length: nsText.length))
            guard !matches.isEmpty else { continue }
            for (offset, match) in matches.enumerated().reversed() {
                let oldValue = nsText.substring(with: match.range)
                let quote = oldValue.contains("\"") ? "\"" : "'"
                let expected = "playOrder=\(quote)\(offset + 1)\(quote)"
                if oldValue != expected {
                    let range = Range(match.range, in: text)!
                    text.replaceSubrange(range, with: expected)
                    repaired += 1
                }
            }
            try text.write(to: file, atomically: true, encoding: .utf8)
        }
        return repaired
    }

    private func isValidXMLIdentifier(_ value: String) -> Bool {
        guard let first = value.unicodeScalars.first,
              first == "_" || CharacterSet.letters.contains(first) else { return false }
        let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "._-"))
        return value.unicodeScalars.dropFirst().allSatisfy { allowed.contains($0) }
    }

    private func sanitizedIdentifier(_ value: String) -> String {
        let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "._-"))
        let body = String(value.unicodeScalars.map { allowed.contains($0) ? Character(String($0)) : "-" })
        return "mv-\(body.isEmpty ? "id" : body)"
    }

    private func containsMissingURL(in text: String, relativeTo file: URL, root: URL) -> Bool {
        !missingURLTokens(in: text, relativeTo: file, root: root).isEmpty
    }

    private func missingURLTokens(in text: String, relativeTo file: URL, root: URL) -> [String] {
        guard let regex = try? NSRegularExpression(
            pattern: #"url\(\s*([\"']?)([^\"')]+)\1\s*\)"#,
            options: [.caseInsensitive]
        ) else { return [] }
        let nsText = text as NSString
        return regex.matches(in: text, range: NSRange(location: 0, length: nsText.length)).compactMap { match in
            let token = nsText.substring(with: match.range)
            let value = nsText.substring(with: match.range(at: 2)).trimmingCharacters(in: .whitespacesAndNewlines)
            return isMissingLocalResource(value, relativeTo: file, root: root) ? token : nil
        }
    }

    private func isMissingLocalResource(_ rawValue: String, relativeTo file: URL, root: URL) -> Bool {
        let lower = rawValue.lowercased()
        if lower.hasPrefix("data:") || lower.hasPrefix("http:") || lower.hasPrefix("https:") || rawValue.hasPrefix("#") {
            return false
        }
        let path = rawValue.split(separator: "#", maxSplits: 1).first.map(String.init) ?? rawValue
        let decoded = path.removingPercentEncoding ?? path
        let candidate = file.deletingLastPathComponent().appendingPathComponent(decoded).standardizedFileURL
        let rootPath = root.standardizedFileURL.path + "/"
        guard candidate.path.hasPrefix(rootPath) else { return true }
        return !FileManager.default.fileExists(atPath: candidate.path)
    }

    private func attribute(named name: String, in tag: String) -> String? {
        guard let regex = try? NSRegularExpression(
            pattern: "\\b\(NSRegularExpression.escapedPattern(for: name))\\s*=\\s*([\\\"'])([^\\\"']+)\\1",
            options: [.caseInsensitive]
        ) else { return nil }
        let nsTag = tag as NSString
        guard let match = regex.firstMatch(in: tag, range: NSRange(location: 0, length: nsTag.length)) else { return nil }
        return nsTag.substring(with: match.range(at: 2))
    }
}

File: Windows/src/MobiVerse.Core/TextEpubPostProcessor.cs
using System.IO.Compression;
using System.Text;
using System.Text.RegularExpressions;

namespace MobiVerse.Core;

public sealed record TextEpubPostProcessResult(
    int RepairedIdentifierCount,
    int RemovedBrokenResourceCount,
    int ReorderedNavigationCount)
{
    public string ReportText => $"""
Text EPUB post-processing
-------------------------
EPUB 3 semantic markup preserved: yes
Invalid identifiers repaired: {RepairedIdentifierCount}
Broken style/resource references removed: {RemovedBrokenResourceCount}
Navigation playOrder values repaired: {ReorderedNavigationCount}
""";
}

public sealed class TextEpubPostProcessor
{
    private static readonly HashSet<string> MarkupExtensions = new(StringComparer.OrdinalIgnoreCase)
    { ".xhtml", ".html", ".htm", ".xml", ".opf", ".ncx" };

    public TextEpubPostProcessResult Process(string epubPath)
    {
        var workspace = TemporaryDirectory();
        try
        {
            EpubPathSecurity.ExtractSafely(epubPath, workspace);
            var files = Directory.EnumerateFiles(workspace, "*", SearchOption.AllDirectories).ToArray();
            var identifiers = RepairInvalidIdentifiers(files);
            var resources = RemoveBrokenResourceReferences(files, workspace);
            var navigation = RepairNavigationOrder(files);
            var result = new TextEpubPostProcessResult(identifiers, resources, navigation);
            if (identifiers + resources + navigation > 0) Rebuild(workspace, epubPath);
            return result;
        }
        finally { try { Directory.Delete(workspace, true); } catch { } }
    }

    private static int RepairInvalidIdentifiers(IReadOnlyList<string> files)
    {
        var regex = new Regex("\\bid\\s*=\\s*([\\\"'])([^\\\"']+)\\1", RegexOptions.IgnoreCase);
        var mapping = new Dictionary<string, string>(StringComparer.Ordinal);
        var used = new HashSet<string>(StringComparer.Ordinal);
        foreach (var file in files.Where(file => MarkupExtensions.Contains(Path.GetExtension(file))))
        {
            foreach (Match match in regex.Matches(File.ReadAllText(file)))
            {
                var identifier = match.Groups[2].Value;
                used.Add(identifier);
                if (IsValidXmlIdentifier(identifier) || mapping.ContainsKey(identifier)) continue;
                var baseValue = SanitizeIdentifier(identifier);
                var replacement = baseValue;
                var suffix = 2;
                while (used.Contains(replacement)) replacement = $"{baseValue}-{suffix++}";
                mapping[identifier] = replacement;
                used.Add(replacement);
            }
        }

        if (mapping.Count == 0) return 0;
        foreach (var file in files.Where(file => MarkupExtensions.Contains(Path.GetExtension(file))))
        {
            var text = File.ReadAllText(file);
            var original = text;
            foreach (var pair in mapping)
            {
                var escaped = Regex.Escape(pair.Key);
                text = Regex.Replace(text, $"(\\bid\\s*=\\s*[\\\"']){escaped}([\\\"'])", $"$1{pair.Value}$2", RegexOptions.IgnoreCase);
                text = text.Replace($"#{pair.Key}", $"#{pair.Value}", StringComparison.Ordinal);
            }
            if (text != original) WriteText(file, text);
        }
        return mapping.Count;
    }

    private static int RemoveBrokenResourceReferences(IReadOnlyList<string> files, string root)
    {
        var removed = 0;
        foreach (var file in files.Where(file => Path.GetExtension(file).Equals(".css", StringComparison.OrdinalIgnoreCase)))
        {
            var css = File.ReadAllText(file);
            var original = css;
            foreach (Match match in Regex.Matches(css, @"@font-face\s*\{.*?\}", RegexOptions.IgnoreCase | RegexOptions.Singleline).Cast<Match>().Reverse())
            {
                if (!MissingUrlTokens(match.Value, file, root).Any()) continue;
                css = css.Remove(match.Index, match.Length);
                removed++;
            }

            foreach (var token in MissingUrlTokens(css, file, root))
            {
                var declaration = new Regex(@"(?im)^\s*[-a-z0-9_]+\s*:\s*[^;{}]*" + Regex.Escape(token) + @"[^;{}]*;\s*");
                var replaced = declaration.Replace(css, "");
                if (replaced != css) { css = replaced; removed++; }
            }
            if (css != original) WriteText(file, css);
        }

        foreach (var file in files.Where(file => new[] { ".xhtml", ".html", ".htm" }.Contains(Path.GetExtension(file), StringComparer.OrdinalIgnoreCase)))
        {
            var markup = File.ReadAllText(file);
            var original = markup;
            foreach (Match match in Regex.Matches(markup, @"<link\b[^>]*>", RegexOptions.IgnoreCase).Cast<Match>().Reverse())
            {
                if (!match.Value.Contains("stylesheet", StringComparison.OrdinalIgnoreCase)) continue;
                var href = Attribute("href", match.Value);
                if (href is null || !IsMissingLocalResource(href, file, root)) continue;
                markup = markup.Remove(match.Index, match.Length);
                removed++;
            }
            if (markup != original) WriteText(file, markup);
        }
        return removed;
    }

    private static int RepairNavigationOrder(IReadOnlyList<string> files)
    {
        var repaired = 0;
        var regex = new Regex("playOrder\\s*=\\s*([\\\"'])\\d+\\1", RegexOptions.IgnoreCase);
        foreach (var file in files.Where(file => Path.GetExtension(file).Equals(".ncx", StringComparison.OrdinalIgnoreCase)))
        {
            var index = 0;
            var text = File.ReadAllText(file);
            var replaced = regex.Replace(text, match =>
            {
                index++;
                var expected = $"playOrder={match.Groups[1].Value}{index}{match.Groups[1].Value}";
                if (match.Value != expected) repaired++;
                return expected;
            });
            if (replaced != text) WriteText(file, replaced);
        }
        return repaired;
    }

    private static IEnumerable<string> MissingUrlTokens(string text, string file, string root)
    {
        var regex = new Regex("url\\(\\s*([\\\"']?)([^\\\"')]+)\\1\\s*\\)", RegexOptions.IgnoreCase);
        return regex.Matches(text).Cast<Match>()
            .Where(match => IsMissingLocalResource(match.Groups[2].Value.Trim(), file, root))
            .Select(match => match.Value).ToArray();
    }

    private static bool IsMissingLocalResource(string rawValue, string file, string root)
    {
        if (rawValue.StartsWith("data:", StringComparison.OrdinalIgnoreCase) ||
            rawValue.StartsWith("http:", StringComparison.OrdinalIgnoreCase) ||
            rawValue.StartsWith("https:", StringComparison.OrdinalIgnoreCase) || rawValue.StartsWith('#')) return false;
        var value = rawValue.Split('#', 2)[0];
        try { value = Uri.UnescapeDataString(value); } catch { }
        var candidate = Path.GetFullPath(Path.Combine(Path.GetDirectoryName(file)!, value.Replace('/', Path.DirectorySeparatorChar)));
        return !EpubPathSecurity.Contains(candidate, root) || !File.Exists(candidate);
    }

    private static string? Attribute(string name, string tag)
    {
        var match = Regex.Match(tag, $"\\b{Regex.Escape(name)}\\s*=\\s*([\\\"'])([^\\\"']+)\\1", RegexOptions.IgnoreCase);
        return match.Success ? match.Groups[2].Value : null;
    }

    private static bool IsValidXmlIdentifier(string value) =>
        value.Length > 0 && (value[0] == '_' || char.IsLetter(value[0])) && value.Skip(1).All(character => char.IsLetterOrDigit(character) || ".-_".Contains(character));

    private static string SanitizeIdentifier(string value)
    {
        var body = new string(value.Select(character => char.IsLetterOrDigit(character) || ".-_".Contains(character) ? character : '-').ToArray());
        return "mv-" + (body.Length == 0 ? "id" : body);
    }

    private static void Rebuild(string root, string epubPath)
    {
        var replacement = Path.Combine(Path.GetDirectoryName(Path.GetFullPath(epubPath))!, $".mobiverse-text-repair-{Guid.NewGuid():N}.epub");
        try
        {
            using (var stream = new FileStream(replacement, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None))
            using (var archive = new ZipArchive(stream, ZipArchiveMode.Create))
            {
                var mimetype = Path.Combine(root, "mimetype");
                if (File.Exists(mimetype)) AddFile(archive, mimetype, "mimetype", CompressionLevel.NoCompression);
                foreach (var file in Directory.EnumerateFiles(root, "*", SearchOption.AllDirectories)
                             .Where(file => !file.Equals(mimetype, StringComparison.OrdinalIgnoreCase))
                             .OrderBy(file => Path.GetRelativePath(root, file), StringComparer.Ordinal))
                    AddFile(archive, file, Path.GetRelativePath(root, file).Replace('\\', '/'), CompressionLevel.Optimal);
            }
            File.Move(replacement, epubPath, true);
        }
        finally { try { File.Delete(replacement); } catch { } }
    }

    private static void AddFile(ZipArchive archive, string path, string name, CompressionLevel level)
    {
        var entry = archive.CreateEntry(name, level);
        entry.LastWriteTime = new DateTimeOffset(1980, 1, 1, 0, 0, 0, TimeSpan.Zero);
        using var input = File.OpenRead(path);
        using var output = entry.Open();
        input.CopyTo(output);
    }

    private static void WriteText(string path, string value) => File.WriteAllText(path, value, new UTF8Encoding(false));

    private static string TemporaryDirectory()
    {
        var path = Path.Combine(Path.GetTempPath(), "MobiVerseTextEpubRepair", Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(path);
        return path;
    }
}
